### PR TITLE
New Audit: using deprecated APIs

### PR DESCRIPTION
**Pull request by ebidel**
R: all 

Part of #1458.

screen shot 2017-01-13 at 4 56 37 pm

### CODE REVIEW

File: lighthouse-core/gather/gatherers/console-violations.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+/**
+ * @fileoverview Gathers console deprecation and intervention warnings logged by Chrome.
+ */
+
+'use strict';
+
+const Gatherer = require('./gatherer');
+
+class ConsoleViolations extends Gatherer {
```
Comment (paulirish) [inline lighthouse-core/gather/gatherers/console-violations.js]: Afaict, this will gather *all* logs added to console rather than just violations, right?

Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-violations.js]: That's right. All logging generated by chrome. Not those using the `console` api. Are you thinking a rename of the class?

Comment (paulirish) [inline lighthouse-core/gather/gatherers/console-violations.js]: Basically yeah. Console logs don't go through Log.entryAdded? Thought they did!

Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-violations.js]: Apparently they've been broken out into https://chromedevtools.github.io/debugger-protocol-viewer/tot/Runtime/#event-consoleAPICalled

Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-violations.js]: done

File: lighthouse-core/gather/gatherers/console-log-entry.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    return driver.sendCommand('Log.disable').then(_ => this._logEntries);
+  }
+
+  beforePass(options) {
+    return this.beginConsoleMessageCollect(options.driver);
+  }
+
+  afterPass(options) {
+    return this.endConsoleMessageCollect(options.driver)
+        .then(entries => {
```
Comment (patrickhulce) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: nit: remove this function and just go to catch?

File: lighthouse-core/config/default.json [hunk 3]
```diff
@@ -334,6 +336,14 @@
         }
       }
     }, {
+      "name": "Avoiding deprecated APIs and browser interventions",
+      "audits": {
+        "deprecations": {
+          "expectedValue": true,
```
Comment (patrickhulce) [inline lighthouse-core/config/default.json]: Won't these not do anything since it's unscored?

Comment (ebidel) [inline lighthouse-core/config/default.json]: Yep. Just adding as a creative of habit. Removed.

File: lighthouse-core/gather/gatherers/console-log-entry.js (hunk 2 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: done

**Comment by ebidel** [comment]
PTAL

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.css [hunk 4]
```diff
@@ -17,4 +17,4 @@
 
 body {
   background-color: #eee;
-}
+}
```
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.css]: new line was bad?

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html [hunk 5]
```diff
@@ -299,6 +307,24 @@
   stampTemplate('unoptimized-images-tmpl', document.body);
 }
 
+function deprecationsTest() {
+  stampTemplate('deprecations-tmpl', document.head);
+
+  // Some APIs atm are not triggering Chrome's deprecation messags. https://crbug.com/680832.
```
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: `messages`, but maybe something more like `// TODO: some deprecated API messages are not currently being logged correctly. See: https://crbug.com/680832`

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html [hunk 6]
```diff
@@ -299,6 +307,24 @@
...
+  // try {
+  //   MediaStreamTrack.getSources(); // FAIL
+  // } catch(e) {}
+
+  // console.timeline(); // FAIL
+
+  // FAIL
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', 'https://lighthouse-viewer.appspot.com/data', false);
```
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: can you do this to the local server? Ideally we wouldn't have any external connections for `npm run smoke`

File: lighthouse-core/gather/gatherers/console-log-entry.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+/**
+ * @fileoverview Gathers console deprecation and intervention warnings logged by Chrome.
+ */
+
+'use strict';
+
+const Gatherer = require('./gatherer');
+
+class ConsoleLogEntry extends Gatherer {
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: this sounds like "`console.log` entries" now :) What about just `LogEntries` to mirror debugger protocol's `Log` domain? (also made it plural to match e.g. `tags-blocking-first-paint`, `styles`, `all-event-listeners`, etc

File: lighthouse-core/gather/gatherers/console-log-entry.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    super();
+    this._logEntries = [];
+    this._onConsoleEntryAdded = this.onConsoleEntry.bind(this);
+  }
+
+  onConsoleEntry(entry) {
+    this._logEntries.push(entry);
+  }
+
+  beginConsoleMessageCollect(driver) {
```
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: just inline `beginConsoleMessageCollect` and `endConsoleMessageCollect`? They're really short

File: lighthouse-core/audits/deprecations.js [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+/**
+ * @fileoverview Audits a page to determine if it is calling deprecated APIs.
```
Comment (brendankenny) [inline lighthouse-core/audits/deprecations.js]: maybe slightly more specific about auditing the log messages to surface deprecated API warnings or something?

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html [hunk 10]
```diff
@@ -299,6 +307,24 @@
...
 
+function deprecationsTest() {
+  stampTemplate('deprecations-tmpl', document.head);
+
+  // Some APIs atm are not triggering Chrome's deprecation messags. https://crbug.com/680832.
+  // try {
+  //   MediaStreamTrack.getSources(); // FAIL
+  // } catch(e) {}
+
+  // console.timeline(); // FAIL
```
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: would it be crazy to uncomment these so we'd be forced to keep up to date as they're fixed and so the expectations file would no longer match? :)

File: lighthouse-core/gather/gatherers/console-log-entry.js (hunk 7 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: Not a great name, is it? Hmm. We need something in between console messages and console logs entries only produced by chrome. `LogEntries` is better but it's still ambiguous. What about 'ChromeConsoleLogs`?

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.css (hunk 4 above)
Comment (ebidel) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.css]: not sure why vscode stripped it. Done.

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html (hunk 5 above)
Comment (ebidel) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: done

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html (hunk 10 above)
Comment (ebidel) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: I don't think we want to do that b/c smokehouse tests will fail indefinitely until Chrome fixes it. That could be an annoying while :\

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html (hunk 6 above)
Comment (ebidel) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: done.

File: lighthouse-core/gather/gatherers/console-log-entry.js (hunk 8 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: done

File: lighthouse-core/audits/deprecations.js (hunk 9 above)
Comment (ebidel) [inline lighthouse-core/audits/deprecations.js]: done

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html (hunk 10 above)
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: oh, I meant uncomment these but keep smokehouse expectation at 4 and let chrome eventually fixing the broken deprecations break the test, then update the expectations. But no problem

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html [hunk 11]
```diff
@@ -308,6 +316,25 @@
...
+  // See: https://crbug.com/680832
+  // try {
+  //   MediaStreamTrack.getSources(); // FAIL
+  // } catch(e) {}
+
+  // console.timeline(); // FAIL
+
+  // FAIL
+  const xhr = new XMLHttpRequest();
+  xhr.open('GET', 'dbw_tester.html', false);
```
Comment (brendankenny) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: sorry, forgot to include earlier: on the topic of cross origin requests, can you change the `https://unknown.com` (which appears to be a real site, just not https) to something like `https://unknown.invalid`, which is reserved for testing?

File: lighthouse-core/gather/gatherers/console-log-entry.js (hunk 7 above)
Comment (brendankenny) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: I feel like having console + log still associates with `console.log`. `ChromeConsoleMessages`?

File: lighthouse-cli/test/smokehouse/dobetterweb/dbw-expectations.js [hunk 12]
```diff
@@ -117,7 +117,7 @@ module.exports = [
         score: false,
         extendedInfo: {
           value: {
-            length: 4
+            length: 5
```
Comment (brendankenny) [inline lighthouse-cli/test/smokehouse/dobetterweb/dbw-expectations.js]: how did this one change?

File: lighthouse-core/gather/gatherers/console-log-entry.js (hunk 7 above)
Comment (ebidel) [inline lighthouse-core/gather/gatherers/console-log-entry.js]: done

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html (hunk 10 above)
Comment (ebidel) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: Oh got it. Done.

File: lighthouse-cli/test/smokehouse/dobetterweb/dbw-expectations.js (hunk 12 above)
Comment (ebidel) [inline lighthouse-cli/test/smokehouse/dobetterweb/dbw-expectations.js]: The xhr: `xhr.open('GET', 'dbw_tester.html', false);`. Moved from using an http2 endpoint (lighthouse-viewer.appspot.com).

File: lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html (hunk 11 above)
Comment (ebidel) [inline lighthouse-cli/test/fixtures/dobetterweb/dbw_tester.html]: done

**Comment by ebidel** [comment]
PTAL

**Comment by ebidel** [comment]
Travis is happy now.

**Comment by ebidel** [comment]
Horse powerful IOT warning?

**Pull request merged**